Fibonacci and NegaFibonacci Series

prathish-ks opened the pull request:
Fixes #  #N/A

Review of colleague's PR #313 - https://github.com/anz-bank/go-course/pull/313

#### Changes proposed in this PR:

## - Fibonacci and NegaFibonacci Series - Lab1

Commit:
Fibonacci lab-1 completed along with negative series

prathish-ks committed:
Print Fibonacci series upto 7

This program prints fibonacci series upto 7 as per Lab one exercise
requirement. It's also programmed for negative fibonacci series as well,
as tested in main_test. Unit Testing has 100% coverage and the code has
been lint as well.

sriyapaka commented:
### 

> [`****`](

- 

- [ ] @ #1. url

)

joel00wood commented on 01_fib/sriyapaka/main_test.go:
```diff
@@ -0,0 +1,31 @@
+	expected := strconv.Quote("1\n1\n2\n3\n5\n8\n13\n")
+	actual := strconv.Quote(buf.String())
+	if expected != actual {
+		t.Errorf("unexpected result - fib function")
```
It would probably help if you printed the expected and actual outputs here, in the event this error is triggered you won't actually know why. The same goes for when this was used again in `TestNegativeSeries()`.

Further to that it would probably help if you looked at using table driven tests, this will enable you to merge the `TestNegativeSeries` test case into your `TestFib` test case. From the slides, this should give you a good reference on how to approach this: https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

joel00wood commented on 01_fib/sriyapaka/main.go:
```diff
@@ -0,0 +1,43 @@
+			b = c
+		} else {
+			c = a + b
+			fmt.Fprintln(out, c)
```
As this is performed for c in both branches of the if statement you could probably lift this out and have it executed after the if statement has assigned a value to c

joel00wood reviewed (commented):
Couple comments regarding simplifying your tests to a single table driven test. I'd further recommend separating out or being more explicit about testing of the `main()` function, including testing for your `abs(x int)` function as well as providing some more test cases for `fib(n int)` on it's own.

Only remaining thing would be removing the following binary files as they aren't necessary for this PR:
- .DS_Store
- 01_fib/.DS_Store
- 01_fib/sriyapaka/sriyapaka

davidbroughsmyth reviewed (commented):
Suggestion, you may want to add .DS_Store to your .gitignore file so that git will ignore these files

prathish-ks committed:
Resolve PR Review Comments

Removed three unnecessary files from the commit. Table driven testing used
now to test multiple test cases. Test added for abs function. In Main.go,
Common statement in if branches has been lifted out

prathish-ks committed:
Remove files from Pull request

joel00wood commented:
Changes look good to me!

prathish-ks commented:
> Changes look good to me!

Thanks for your time. This is my first ever pull request :)

juliaogris commented on 01_fib/sriyapaka/main.go:
```diff
@@ -0,0 +1,42 @@
+}
+
+func abs(x int) int {
+
```
Unnecessary blank line.
No need for blank lines at beginning or end of block.
Use blank lines between type, method and function definitions and within a block for logical separation.
Please fix throughout this file.

juliaogris commented on 01_fib/sriyapaka/main.go:
```diff
@@ -0,0 +1,42 @@
+	var sign = n / absn
+
+	for i := 0; i < absn; i++ {
+		if i > 0 {
```
It's functionally correct, but could be improved:
Don't use conditions on loop index but rearrange your loop.
Here is what _we_ have come up with:
```go
	sign := // 1 for positive n ; -1 for negative n
	n1, n2 := 0, 1
	for j := 0; j < n*sign; j++ {
		n1, n2 = n2, n1+sign*n2
		// print n1
	}
```

juliaogris commented on 01_fib/sriyapaka/main_test.go:
```diff
@@ -0,0 +1,77 @@
+		out = &buf
+		test := test
+
+		t.Run(name, func(t *testing.T) {
```
Great!

juliaogris reviewed (changes requested):
Looks pretty close!

Please read through the [Collected Code Review Comments](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments)
- [Fix your pr description](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#pr-description)
- [Limit the number of commits to one or two](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#limit-the-number-of-commits-to-one-or-two)

juliaogris commented:
The go-course is now closed. Thank you very much for participating.
